src/awkward/operations/structure.py

```diff
@@ -2157,6 +2157,202 @@ def getfunction(layout, depth, posaxis):
+            layout,
+            (
+                ak.layout.ListArray32,
+                ak.layout.ListArrayU32,
+                ak.layout.ListArray64,
+            ),
+        ):
+            return apply(layout.toListOffsetArray64(True), depth, posaxis)
+
+        # ListOffsetArray performs resizing
+        if isinstance(
+            layout,
+            (
+                ak.layout.ListOffsetArray32,
+                ak.layout.ListOffsetArray64,
+                ak.layout.ListOffsetArrayU32,
+            ),
+        ):
+            new_layout = layout.toListOffsetArray64(True)
+            return ak.layout.ListOffsetArray64(
+                new_layout.offsets,
+                apply(new_layout.content, depth + 1, posaxis),
+                new_layout.identities,
+                new_layout.parameters,
+            )
+
+        # UnmaskedArray just wraps another array
+        if isinstance(layout, ak.layout.UnmaskedArray):
+            return ak.layout.UnmaskedArray(apply(layout.content, depth, posaxis))
+
+        # UnionArrays can be simplified
+        # and their contents too
+        if isinstance(layout, ak._util.uniontypes):
+            layout = layout.simplify()
+
+            # If we managed to lose the drop type entirely
+            if not isinstance(layout, ak._util.uniontypes):
+                return apply(layout, depth, posaxis)
+
+            # Pack simplified layout
+            tags = nplike.asarray(layout.tags)
+            index = nplike.asarray(layout.index)
+
+            new_contents = [None] * len(layout.contents)
+            new_index = nplike.zeros_like(index)
+
+            # Compact indices
+            for i in range(len(layout.contents)):
+                is_i = tags == i
+
+                new_contents[i] = apply(layout.project(i), depth, posaxis)
+                new_index[is_i] = nplike.arange(nplike.sum(is_i))
+
+            return ak.layout.UnionArray8_64(
+                ak.layout.Index8(tags),
+                ak.layout.Index64(new_index),
+                new_contents,
+                layout.identities,
+                layout.parameters,
+            )
+
+        # RecordArray contents can be truncated
+        if isinstance(layout, ak.layout.RecordArray):
+            return ak.layout.RecordArray(
+                [
+                    apply(truncate(c, len(layout)), depth, posaxis)
+                    for c in layout.contents
+                ],
+                layout.recordlookup,
+                len(layout),
+                layout.identities,
+                layout.parameters,
+            )
+
+        # RegularArrays can change length
+        if isinstance(layout, ak.layout.RegularArray):
+            if not len(layout):
+                return layout
+
+            content = layout.content
+
+            # Truncate content if it is larger than a perfect
+            # multiple of the RegularArray size
+            n, r = divmod(len(content), layout.size)
+            if r:
```

Comment:
```python
if r != 0:
```

(I know that it's allowed in Python and black/flake8, but I want to avoid all weak typing, and Python's truthiness is the only part of Python that's weakly typed.)